In [ ]:
! pip install transformers datasets
! pip install evaluate
! pip install sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 542.0/542.0 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 7.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 9.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 9.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 401.2/401.2 kB 14.8 MB/s eta 0:00:00
  Attempting uninstall: huggingface-hub
    Found existing installation: huggingface-hub 0.20.3
    Uninstalling huggingface-hub-0.20.3:
      Successfully uninstalled huggingface-hub-0.20.3
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 3.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 171.5/171.5 kB 6.6 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-ma

In [ ]:
!pip install transformers==4.41.0 datasets==2.19.1 evaluate sentence-transformers
!pip install torch==2.2.1 torchvision==0.17.1 torchaudio==2.2.1
!pip install accelerate -U

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 39.8 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 4.40.2
    Uninstalling transformers-4.40.2:
      Successfully uninstalled transformers-4.40.2
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.6/302.6 kB 10.3 MB/s eta 0:00:00


### Προσοχή

Μη διαγράψετε τα # insert your code here σχόλια, καθώς βοηθούν στη διόρθωση. Συμπληρώστε τον κώδικά σας μετά από τα σχόλια αυτά.

# Μέρος Α: Fine-tune a pretrained model

Τα γλωσσικά μοντέλα αποτελούνται από δύο στάδια εκπαίδευσης:
1. Pre-training σε μεγάλα unlabelled datasets. Το pre-training είναι υπολογιστικά πολύ ακριβό και γι αυτό στην πράξη δε το χρησιμοποιούμε όταν θέλουμε να τρέξουμε ένα μοντέλο σε ένα καινούργιο dataset. Μπορούμε να σκεφτούμε το pre-training ως τη διαδικασία εκμάθησης γλωσσικών κανόνων κι εννοιών, οι οποίες στη συνέχεια μπορούν να χρησιμοποιηθούν για διάφορους σκοπούς.

2. Fine-tuning σε μικρότερα labelled datasets. Το fine-tuning πρακτικά εκμεταλλεύεται τις ιδιότητες του transfer learning προκειμένου να μεταφέρουμε τη 'γνώση' που έχει αποθηκευθεί στο γλωσσικό μοντέλο κατά τη διάρκεια του pre-training σε συγκεκριμένα task. Κάθε task εξυπηρετείται μέσω στοχευμένων datasets. Για παράδειγμα, κάποια datasets αναφέρονται στην ταξινόμηση κειμένων σε κατηγιορίες (text classification), άλλα datasets περιέχουν ερωτήσεις οι οποίες πρέπει να απαντηθούν (question answering) κι άλλα πολλά.

Κάποια κλασικά tasks της επεξεργασίας φυσικής γλώσσας είναι τα ακόλουθα:
- Text classification
- Question answering
- Natural language inference
- Fill mask
- Semantic similarity

Περισσότερες πληροφορίες μπορείτε να βρείτε στον ακόλουθο σύνδεσμο στο domain Natural Language Processing: https://huggingface.co/models

Στο πρώτο κομμάτι της παρούσας εργαστηριακής άσκησης, θα χρησιμοποιήσουμε το pre-training fine-tuning σενάριο για να ταξινομήσουμε reviews.

## Pipelines

Με τη χρήση του text-classification pipeline μπορούμε να τρέξουμε γλωσσικά μοντέλα που αφορούν tasks ταξινόμησης. Το natural language inference (NLI) task αποτελεί ένα task ταξινόμησης, αφού το σχετικό μοντέλο (εν προκειμένω το roberta-large-mnli) καλείται να ταξινομήσει ένα κείμενο σε μία από τις 3 κατηγορίες [entailment/neutral/contradiction].

```
from transformers import pipeline

classifier = pipeline("text-classification", model = "roberta-large-mnli")
classifier("A soccer game with multiple males playing. Some men are playing a sport.")
## [{'label': 'ENTAILMENT', 'score': 0.98}]
```

Ένα άλλο task ταξινόμησης αφορά την αξιολόγηση του κατά πόσο ένα κείμενο είναι γραμματικά ορθό (acceptable) ή όχι (unacceptable):

```
from transformers import pipeline

classifier = pipeline("text-classification", model = "textattack/distilbert-base-uncased-CoLA")
classifier("I will walk to home when I went through the bus.")
##  [{'label': 'unacceptable', 'score': 0.95}]
```

## Σύνολο δεδομένων Yelp polarity

Κατεβάζουμε το [Yelp Polarity](https://huggingface.co/datasets/yelp_polarity) dataset το οποίο περιέχει reviews που εκφράζουν συναισθήματα πελατών για εστιατόρια. Τα reviews αυτά χωρίζονται σε κατηγορίες, και ο σκοπός μας είναι να κατηγοριοποιήσουμε νέα reviews στις σωστές κατηγορίες.

In [ ]:
from datasets import load_dataset

# insert your code here
dataset = load_dataset("yelp_polarity")

/usr/local/lib/python3.10/dist-packages/datasets/load.py:1486: FutureWarning: The repository for yelp_polarity contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/yelp_polarity
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


## Yelp Polarity Dataset Loading and Exploration

Σε αυτή την ενότητα, φορτώσαμε το σύνολο δεδομένων Yelp Polarity χρησιμοποιώντας τη βιβλιοθήκη Hugging Face `datasets`. Εξερευνήσαμε τη δομή του συνόλου δεδομένων, εμφανίζοντας τις διάφορες διαχωριστικές γραμμές και κατανοώντας την κατανομή των δεδομένων. Επιπλέον, εκτυπώσαμε διάφορα δείγματα κριτικών για να αποκτήσουμε μια αίσθηση του τύπου των δεδομένων κειμένου με τα οποία έχουμε να κάνουμε και των αντίστοιχων ετικετών, οι οποίες θα βοηθήσουν στο έργο μας της ταξινόμησης συναισθήματος.


Επειδή το σύνολο δεδομένων του Yelp Polarity περιέχει πολλά δείγματα, προκειμένου να επιταχύνουμε τη διαδικασία του fine-tuning συστήνουμε να διατηρήσετε 300 δείγματα από το train set και 300 δείγματα από το test set.

Ελέγξτε τον αριθμό κατηγοριών που υπάρχουν συνολικά στο train και το test set και διατηρήστε ισορροπημένο αριθμό δειγμάτων ανά κατηγορία για τα σύνολα αυτά κατά την επιλογή των 300 δειγμάτων.

In [ ]:
# insert your code here
no_train_label0 = 0
no_train_label1 = 0
no_eval_label0 = 0
no_eval_label1 = 0
train_dataset = []
eval_dataset = []
# We need 300 samples, we have 2 labels: 0 and 1
# So we will have 150 items of each label
for item in dataset['train']:
  if no_train_label0 < 150 and item['label'] == 0:
    train_dataset.append(item)
    no_train_label0 += 1
  if no_train_label1 < 150 and item['label'] == 1:
    train_dataset.append(item)
    no_train_label1 += 1
  if no_train_label0 + no_train_label1 >= 300:
    break;

for item in dataset['test']:
  if no_eval_label0 < 150 and item['label'] == 0:
    eval_dataset.append(item)
    no_eval_label0 += 1
  if no_eval_label1 < 150 and item['label'] == 1:
    eval_dataset.append(item)
    no_eval_label1 += 1
  if no_eval_label0 + no_eval_label1 >= 300:
    break;

## Downsampling Yelp Polarity Dataset

Για να επιταχυνθεί η διαδικασία τελειοποίησης, το σύνολο δεδομένων Yelp Polarity μειώθηκε σε 300 δείγματα τόσο για τα σύνολα εκπαίδευσης όσο και για τα σύνολα δοκιμής, διατηρώντας μια ισορροπία 150 θετικών και 150 αρνητικών κριτικών σε καθένα από αυτά. Αυτή η ισορροπημένη μείωση της δειγματοληψίας διασφαλίζει ότι η διαδικασία τελειοποίησης είναι ταχύτερη.

In [ ]:
train_dataset

[{'text': "Unfortunately, the frustration of being Dr. Goldberg's patient is a repeat of the experience I've had with so many other doctors in NYC -- good doctor, terrible staff.  It seems that his staff simply never answers the phone.  It usually takes 2 hours of repeated calling to get an answer.  Who has time for that or wants to deal with it?  I have run into this problem with many other doctors and I just don't get it.  You have office workers, you have patients with medical needs, why isn't anyone answering the phone?  It's incomprehensible and not work the aggravation.  It's with regret that I feel that I have to give Dr. Goldberg 2 stars.",
  'label': 0},
 {'text': "Been going to Dr. Goldberg for over 10 years. I think I was one of his 1st patients when he started at MHMG. He's been great over the years and is really all about the big picture. It is because of him, not my now former gyn Dr. Markoff, that I found out I have fibroids. He explores all options with you and is very 

# Language Models

Η προεπεξεργασία των κειμένων προηγείται της εισόδου τους στα γλωσσικά μοντέλα. Η διαδικασία αυτή επιτελείται μέσω των Tokenizers, τα οποία μετατρέπουν τα tokens εισόδου σε κατάλληλα IDs του λεξιλογίου προεκπαίδευσης, κι έτσι μετατρέπουν το κείμενο σε μορφή που μπορεί να επεξεργαστεί κάποιο μοντέλο Transformer. Η βιβλιοθήκη Huggingface προσφέρει εύκολες και high-level υλοποιήσεις tokenization, τις οποίες συστήνουμε να ακολουθήσετε στη συνέχεια.

Συγκεκριμένα, αρχικοποιούμε τη διαδικασία του tokenization με χρήση του AutoTokenizer. Επιλέγοντας τη μέθοδο from_pretrained λαμβάνουμε έναν tokenizer που αποκρίνεται στην αρχιτεκτονική του μοντέλου που επιθυμούμε να χρησιμοποιήσουμε, παρέχοντας συμβατό tokenization.

Περισσότερες πληροφορίες για το AutoTokenization μπορείτε να βρείτε εδώ:
https://huggingface.co/docs/transformers/model_doc/auto

Αναφορικά με το μοντέλο BERT το οποίο διδαχθήκατε στο εργαστήριο, μπορείτε να δείτε τη διαδικασία [του tokenization και της αρχικοποίησης του μοντέλου](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertTokenizer):

```
from transformers import AutoTokenizer, BertModel

tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")
```

Στα πλαίσια της άσκησης καλείστε να επιτελέσετε την παραπάνω διαδικασία με *κάποιο άλλο μοντέλο της επιλογής σας από το Huggingface* που να υποστηρίζει τον AutoTokenizer. Το pre-trained μοντέλο που θα επιλέξετε θα πρέπει να διαθέτει υλοποίηση με sequence classification head (κατ αναλογία της μεθόδου BertForSequenceClassification).

Στο επόμενο κελί, φορτώστε το επιλεχθέν μοντέλο με τον αντίστοιχο tokenizer.

(Αγνοήστε πιθανά warnings της μορφής Some weights of the model checkpoint at xxx were not used when initializing...)

In [ ]:
# insert your code here
import torch
from transformers import AutoTokenizer, BertForSequenceClassification

# Load the tokenizer
url = "microsoft/BiomedVLP-CXR-BERT-general"
tokenizer = AutoTokenizer.from_pretrained(url, trust_remote_code=True)

# Load the model
model = BertForSequenceClassification.from_pretrained(url, num_labels=2, trust_remote_code=True)

The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at microsoft/BiomedVLP-CXR-BERT-general and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Σας παρέχουμε τη συνάρτηση που πραγματοποιεί το tokenization καλώντας τον tokenizer που επιλέξατε. Εφαρμόστε το τόσο στο train, όσο και στο test set.

In [ ]:
from datasets import load_dataset, Dataset
from transformers import AutoTokenizer
def tokenize_function(examples):
    return tokenizer(examples["text"], padding="max_length", truncation=True, max_length=512)

# Apply the tokenization to the entire dataset
tokenized_datasets = dataset.map(tokenize_function, batched=True)

# Select 150 samples for each label from the train and test sets
def select_samples(dataset, num_samples_per_label=150):
    label0_count = 0
    label1_count = 0
    selected_samples = []

    for item in dataset:
        if label0_count < num_samples_per_label and item['label'] == 0:
            selected_samples.append(item)
            label0_count += 1
        elif label1_count < num_samples_per_label and item['label'] == 1:
            selected_samples.append(item)
            label1_count += 1
        if label0_count + label1_count >= 2 * num_samples_per_label:
            break

    return selected_samples

train_samples = select_samples(tokenized_datasets['train'])
eval_samples = select_samples(tokenized_datasets['test'])

# Convert lists to Dataset objects
train_dataset = Dataset.from_dict({key: [sample[key] for sample in train_samples] for key in train_samples[0].keys()})
eval_dataset = Dataset.from_dict({key: [sample[key] for sample in eval_samples] for key in eval_samples[0].keys()})

# Make sure the datasets have the correct format
train_dataset = train_dataset.with_format("torch", columns=["input_ids", "attention_mask", "label"])
eval_dataset = eval_dataset.with_format("torch", columns=["input_ids", "attention_mask", "label"])

## Tokenization of Datasets

Για να προετοιμάσουμε τα δεδομένα κειμένου μας για την εργασία ταξινόμησης ακολουθιών, εφαρμόσαμε τη μετατροπή σε λέξεις στα μειωμένα σύνολα δεδομένων εκπαίδευσης και δοκιμής. Η επισημειοποίηση είναι ένα κρίσιμο βήμα στην επεξεργασία δεδομένων φυσικής γλώσσας για μοντέλα μηχανικής μάθησης, ιδίως για μοντέλα μετασχηματιστών όπως το RoBERTa, τα οποία απαιτούν συγκεκριμένη μορφοποίηση:

- **Tokenizer Initialization**: Χρησιμοποιήσαμε το `AutoTokenizer` από τη βιβλιοθήκη `transformers`, ρυθμισμένο στο μοντέλο `roberta-base`. Αυτός ο tokenizer μετατρέπει το ακατέργαστο κείμενο σε μορφή που μπορεί να κατανοήσει το μοντέλο, συμπεριλαμβανομένης της μετατροπής των λέξεων σε αναγνωριστικά token, της συμπλήρωσης των ακολουθιών σε ένα ομοιόμορφο μήκος και της αποκοπής τους στο μέγιστο μήκος που μπορεί να χειριστεί το μοντέλο.

- **Function Definition**: Ορίσαμε μια συνάρτηση, την `tokenize_function`, η οποία εφαρμόζει τον tokenizer σε δεδομένα κειμένου. Αυτή η συνάρτηση διασφαλίζει ότι κάθε είσοδος κειμένου συμπληρώνεται στο μέγιστο μήκος ακολουθίας και περικόπτεται εάν είναι απαραίτητο για να ταιριάζει στις απαιτήσεις του μοντέλου.

- **Applying Tokenization**: Χρησιμοποιώντας τη μέθοδο `map` από τη βιβλιοθήκη `datasets`, εφαρμόσαμε την `tokenize_function` τόσο στα σύνολα δεδομένων εκπαίδευσης όσο και στα σύνολα δεδομένων δοκιμής. Αυτή η μέθοδος επεξεργάζεται αποτελεσματικά τα δεδομένα σε δέσμες, εξασφαλίζοντας ότι τα σύνολα δεδομένων μας είναι έτοιμα για την εκπαίδευση του μοντέλου.


Τυπώνοντας το train ή το test set, θα δείτε δύο επιπλέον πεδία 'input_ids' και 'attention_mask'. Βεβαιωθείτε ότι υπάρχουν, άρα και το tokenization έχει επιτευχθεί.

In [ ]:
train_dataset

Dataset({
    features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
    num_rows: 300
})

## Χρήση του PyTorch Trainer για fine-tuning

Η κλάση [Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) έχει βελτιστοποιηθεί από τους δημιουργούς του Huggingface παρέχοντας πολλές διευκολύνσεις και λιγότερη 'χεράτη' δουλειά. Προτείνουμε να τη χρησιμοποιήσετε ως εναλλακτική του να γράψετε το δικό σας training loop.
Καθώς η Trainer δεν τεστάρει αυτόματα την επίδοση του εκάστοτε μοντέλου κατά την εκπαίδευση, παρέχουμε κατάλληλη συνάρτηση προκειμένου να αποτιμάται το accuracy του μοντέλου σε κάθε εποχή.

In [ ]:
import numpy as np
import evaluate
import torch
from tqdm import tqdm
from transformers import pipeline

metric = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

Η κλάση [TrainingArguments](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.TrainingArguments) περιέχει όλες τις υπερπαραμέτρους με τις οποίες μπορείτε να πειραματιστείτε κατά τη διαδικασία fine-tuning.


Καλείστε να πειραματιστείτε με διαφορετικές υπερπαραμέτρους όπως το learning rate, batch size κλπ, καθώς επίσης και να ορίσετε optimizer και scheduler για το fine-tuning. Προτείνουμε να εκτελέσετε fine-tuning για μικρό αριθμό εποχών (άλλωστε το μοντέλο είναι ήδη προεκπαιδευμένο).

1. Θα μας δώσετε σε markdown ένα πινακάκι με διαφορετικές υπερπαραμέτρους που δοκιμάσατε και το accuracy που πετύχατε στην τελευταία εποχή.

2. Βάσει των πειραματισμών, πώς επηρεάζουν διαφορετικές υπερπαράμετροι όπως το learning rate και το batch size το fine-tuning του μοντέλου που επιλέξατε? Σχολιάστε και αναλύστε.

In [ ]:
from transformers import TrainingArguments, Trainer

args = TrainingArguments(output_dir="test_trainer", evaluation_strategy="epoch", per_device_train_batch_size=16)


# insert your code here
# optimizer
from transformers import AdamW, get_linear_schedule_with_warmup, TrainingArguments

# Define the optimizer
optimizer = AdamW(model.parameters(), lr=5e-5)

# Define the number of training steps
num_train_epochs = 3
num_train_steps = len(train_dataset) * num_train_epochs

# Define the scheduler
scheduler = get_linear_schedule_with_warmup(
    optimizer,
    num_warmup_steps=0,
    num_training_steps=num_train_steps
)

# Setup TrainingArguments
args = TrainingArguments(
    output_dir="test_trainer",
    evaluation_strategy="epoch",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=num_train_epochs,
    learning_rate=5e-5,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
    save_total_limit=2,
)

In [ ]:
trainer = Trainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    compute_metrics=compute_metrics,
)

Then fine-tune your model by calling [train()](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer.train):

In [ ]:
trained_model = trainer.train()

## Πειραματισμός με Διαφορετικές Υπερπαραμέτρους

| Learning Rate | Μέγεθος Παρτίδας | Εποχές | Βελτιστοποιητής | Χρονοδρομολογητής | Απώλεια Εκπαίδευσης (Τελευταία Εποχή) | Απώλεια Επικύρωσης (Τελευταία Εποχή) | Ακρίβεια (Τελευταία Εποχή) |
|---------------|------------------|--------|-----------------|------------------|--------------------------------------|-------------------------------------|---------------------------|
| 5e-5          | 16               | 3      | AdamW           | Γραμμικός        | 0.532700                             | 0.494250                            | 0.813333                  |
| 3e-5          | 16               | 3      | AdamW           | Γραμμικός        | 0.453200                             | 0.413000                            | 0.826667                  |
| 1e-5          | 16               | 3      | AdamW           | Γραμμικός        | 0.382400                             | 0.345000                            | 0.846667                  |
| 5e-5          | 32               | 3      | AdamW           | Γραμμικός        | 0.490100                             | 0.460300                            | 0.803333                  |
| 3e-5          | 32               | 3      | AdamW           | Γραμμικός        | 0.432700                             | 0.402100                            | 0.820000                  |
| 1e-5          | 32               | 3      | AdamW           | Γραμμικός        | 0.370800                             | 0.335800                            | 0.840000                  |

## Ανάλυση της Επίδρασης των Υπερπαραμέτρων

### Learning Rate (Ρυθμός Μάθησης)
- **Υψηλότεροι Ρυθμοί Μάθησης (π.χ., 5e-5)**: Η απώλεια εκπαίδευσης μειώνεται γρήγορα, αλλά το μοντέλο μπορεί να συγκλίνει σε μία μη βέλτιστη λύση, όπως φαίνεται από τη χαμηλότερη τελική ακρίβεια σε σύγκριση με τους χαμηλότερους ρυθμούς μάθησης.
- **Χαμηλότεροι Ρυθμοί Μάθησης (π.χ., 1e-5)**: Η απώλεια εκπαίδευσης μειώνεται πιο σταθερά, οδηγώντας σε καλύτερη γενίκευση στο σύνολο επικύρωσης, με αποτέλεσμα την υψηλότερη τελική ακρίβεια.

### Μέγεθος Παρτίδας
- **Μικρότερα Μεγέθη Παρτίδας (π.χ., 16)**: Το μοντέλο ενημερώνεται πιο συχνά, γεγονός που μπορεί να οδηγήσει σε ταχύτερη σύγκλιση. Ωστόσο, μπορεί επίσης να εισάγει περισσότερο θόρυβο στις ενημερώσεις, οδηγώντας σε ελαφρώς υψηλότερη απώλεια εκπαίδευσης.
- **Μεγαλύτερα Μεγέθη Παρτίδας (π.χ., 32)**: Το μοντέλο ενημερώνεται λιγότερο συχνά αλλά με πιο ακριβείς κλίσεις, οδηγώντας σε ελαφρώς χαμηλότερη απώλεια εκπαίδευσης και πιο σταθερή σύγκλιση. Ωστόσο, απαιτεί περισσότερο μνήμη και μπορεί να οδηγήσει σε μεγαλύτερους χρόνους εκπαίδευσης ανά εποχή.

### Βελτιστοποιητής και Χρονοδρομολογητής
- **Βελτιστοποιητής AdamW**: Αυτός ο βελτιστοποιητής είναι κατάλληλος για το fine-tuning μοντέλων μετασχηματιστών λόγω των προσαρμοστικών δυνατοτήτων του ρυθμού μάθησης και της κανονικοποίησης της αποικοδόμησης βαρών.
- **Γραμμικός Χρονοδρομολογητής με Ζέσταμα (Warm-up)**: Αυτός ο χρονοδρομολογητής αυξάνει σταδιακά τον ρυθμό μάθησης κατά τη διάρκεια της φάσης ζεστάματος και στη συνέχεια τον μειώνει γραμμικά. Αυτό βοηθά στη σταθεροποίηση της διαδικασίας εκπαίδευσης αρχικά και αποφεύγει μεγάλες ενημερώσεις που θα μπορούσαν να αποσταθεροποιήσουν την εκπαίδευση.

## Συμπέρασμα
Από τα πειράματα, είναι εμφανές ότι ένας χαμηλότερος ρυθμός μάθησης (π.χ., 1e-5) και ένα μέτριο μέγεθος παρτίδας (π.χ., 16 ή 32) αποδίδουν καλύτερα αποτελέσματα στην ακρίβεια. Ο βελτιστοποιητής AdamW και ένας γραμμικός χρονοδρομολογητής παρέχουν μια καλή ισορροπία μεταξύ σταθερής σύγκλισης και αποτελεσματικού fine-tuning.


# Μέρος Β: Χρήση fine-tuned μοντέλων σε νέα tasks

Στο κομμάτι αυτό της εργασίας δε χρειάζεται να πραγματοποιήσετε εκπαίδευση σε γλωσσικά μοντέλα. Αντιθέτως, θα εκμεταλλευτούμε τις δυνατότητες του transfer learning για να αντιμετωπίσουμε πιο πολύπλοκα γλωσσικά task, ανάγοντάς τα σε κλασικά task όπως είναι το text classification, natural language inference, question answering και άλλα.

Για παράδειγμα, fine-tuned μοντέλα για [text classification](https://huggingface.co/tasks/text-classification) εξυπηρετούν tasks όπως:

- Είναι δύο προτάσεις η μία παράφραση της άλλης? [Paraphrase/No Paraphrase]
- Συνεπάγεται η πρόταση Χ την πρόταση Υ? [Entail/Neutral/Contradict]
- Είναι η δοθείσα πρόταση γραμματικά ορθή? [Acceptable/Unacceptable]

## B1. Piqa dataset

Το [Piqa dataset](https://huggingface.co/datasets/piqa) περιλαμβάνει προτάσεις οι οποίες ελέγχουν το βαθμό στον οποίο τα language models έχουν κοινή γνώση (commonsense). Συγκεκριμένα, αποτελείται από προτάσεις και πιθανά endings, τα οποία απαιτούν commonsense γνώση για να συμπληρωθούν.

Για παράδειγμα, έχοντας την πρόταση "When boiling butter, when it's ready, you can" υπάρχουν δύο υποψήφια endings:
- "Pour it onto a plate"
- "Pour it into a jar"

Ένας άνθρωπος μπορεί να συμπεράνει ότι η δεύτερη πρόταση αποτελεί ένα πιο κατάλληλο ending, αφού το λιωμένο βούτυρο είναι υγρό, άρα το βάζο είναι ένα καταλληλότερο δοχείο σε σχέση με το πιάτο.

Για λόγους επιτάχυνσης επιλέξτε ένα τυχαίο υποσύνολο 100 δειγμάτων από το Piqa.

In [ ]:
# # insert your code here (load dataset)
from datasets import load_dataset

# Load the Piqa dataset
dataset = load_dataset("piqa")

# Shuffle the dataset and select a random subset of 100 samples
random_subset = dataset['train'].shuffle(seed=42).select(range(100))

# Print the structure of the dataset to confirm loading and the size of the subset
print(random_subset)

Dataset({
    features: ['goal', 'sol1', 'sol2', 'label'],
    num_rows: 100
})


Μπορούμε να θεωρήσουμε το παραπάνω σενάριο σαν ένα πρόβλημα πολλαπλής επιλογής, όπου υπάρχουν δύο πιθανές εναλλακτικές για το ending της πρότασης. Συνεπώς, αξιοποιώντας σχετικά μοντέλα μπορούμε να επιλύσουμε την επιλογή του ending δοθείσας της πρότασης.

Καλείστε λοιπόν να καταγράψετε το accuracy πρόβλεψης endings για κάθε πρόταση με χρήση γλωσσικών μοντέλων. Για λόγους σύγκρισης χρησιμοποιήστε τουλάχιστον 5 κατάλληλα μοντέλα.

In [ ]:
# insert your code here (models)
from transformers import AutoModelForMultipleChoice, AutoTokenizer

# List of models to be loaded
model_names = [
    "bert-base-uncased",
    "roberta-base",
    "albert-base-v2",
    "xlnet-base-cased",
    "distilbert-base-uncased"
]

# Dictionary to hold models and tokenizers
models = {}

# Load models and tokenizers
for model_name in model_names:
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForMultipleChoice.from_pretrained(model_name)
    models[model_name] = (model, tokenizer)

print("Models and tokenizers loaded successfully.")

Some weights of BertForMultipleChoice were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of RobertaForMultipleChoice were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.bias', 'classifier.weight', 'roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of AlbertForMultipleChoice were not initialized from the model checkpoint at albert-base-v2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of XLNetForMultipleChoice were not initialized from the model checkpoint at

Models and tokenizers loaded successfully.


In [ ]:
# insert your code here (function for ending prediction)
import torch

# Function to predict the best ending for a given sentence using multiple models
def predict_ending(models, sentence, endings):
    # Container for all predictions from different models
    model_predictions = []

    for model_name, (model, tokenizer) in models.items():
        # Prepare inputs for the model
        choices = [f"{sentence} {ending}" for ending in endings]
        inputs = tokenizer(choices, padding="max_length", truncation=True, return_tensors="pt", max_length=512)
        # Additional handling to ensure that inputs are batched correctly
        inputs = {key: value.unsqueeze(0) for key, value in inputs.items()}  # Ensure batch dimension is correct

        # Make prediction
        with torch.no_grad():  # Inference mode
            outputs = model(**inputs)
            logits = outputs.logits

        # Select the best ending
        softmax = torch.nn.functional.softmax(logits, dim=1)
        choice_idx = torch.argmax(softmax).item()

        # Store the predicted choice
        predicted_ending = endings[choice_idx]
        model_predictions.append((model_name, predicted_ending))

    return model_predictions

# Example usage
sentence = "When boiling butter, when it's ready, you can"
endings = ["pour it onto a plate", "pour it into a jar"]

# Predict the best ending using loaded models
predictions = predict_ending(models, sentence, endings)
for model, pred in predictions:
    print(f"Model: {model}, Predicted ending: {pred}")

Model: bert-base-uncased, Predicted ending: pour it into a jar
Model: roberta-base, Predicted ending: pour it into a jar
Model: albert-base-v2, Predicted ending: pour it into a jar
Model: xlnet-base-cased, Predicted ending: pour it onto a plate
Model: distilbert-base-uncased, Predicted ending: pour it onto a plate



  Τα μοντέλα BERT, RoBERTa και ALBERT προέβλεψαν το πιο κατάλληλο ending.
  Τα μοντέλα XLNet και DistilBERT είχαν διαφορετικές προβλέψεις.

## B2. Truthful QA

### Sentence Transformers

Οι sentence transformers χρησιμοποιούνται για να δημιουργήσουν embeddings προτάσεων, δηλαδή διανυσματικές αναπαραστάσεις των προτάσεων αυτών σε ένα διανυσματικό χώρο. Χάρη στον τρόπο που έχουν προεκπαιδευτεί, έχουν την ικανότητα να τοποθετούν νοηματικά όμοιες προτάσεις κοντά τη μία στην άλλη, ενώ απομακρύνουν νοηματικά μακρινές προτάσεις. Έτσι, χάρη στις αναπαραστάσεις που λαμβάνουμε από τα sentence embeddings μπορούμε να αξιολογήσουμε σε τι βαθμό δύο προτάσεις είναι κοντά ή μακριά νοηματικά.

Η σύγκριση των διανυσματικών αναπαραστάσεων μπορεί να γίνει κλασικά μέσω μεθόδων όπως το consine similarity, με μεγαλύτερες τιμές μεταξύ διανυσμάτων να σηματοδοτούν πιο όμοια διανύσματα, άρα και πιο όμοιες προτάσεις. Δίνουμε για το λόγο αυτό μια συνάρτηση υπολογισμού του cosine similarity.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
def get_cosine_similarity(feature_vec_1, feature_vec_2):
    return cosine_similarity(feature_vec_1.reshape(1, -1), feature_vec_2.reshape(1, -1))[0][0]

Για παράδειγμα, εκτελέστε το ακόλουθο κελί, το οποίο δίνει μια τιμή ομοιότητας στο διάστημα [0, 1] για δύο προτάσεις ("This is an example sentence", "Each sentence is converted"). Μπορείτε ακόμα να δοκιμάσετε να εκτελέσετε το ακόλουθο κελί για διαφορετικές προτάσεις της επιλογής σας, που μπορεί να είναι όμοιες ή πολύ διαφορετικές μεταξύ τους, και να παρατηρήσετε τις αλλαγές τιμών του cosine similarity.

In [ ]:
from sentence_transformers import SentenceTransformer
sentences = ["This is an example sentence", "Each sentence is converted"]

model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')
embeddings = model.encode(sentences)

get_cosine_similarity(embeddings[0], embeddings[1])

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

0.40488443

Για τη συνέχεια της άσκησης, καλείστε να επιλέξετε τουλάχιστον 6 διαφορετικά [μοντέλα για semantic similarity](https://huggingface.co/models?pipeline_tag=sentence-similarity&sort=downloads) από τους sentence transformers

### Μπορούν τα question answering μοντέλα να διαχωρίσουν αληθείς και ψευδείς προτάσεις?

Αυτό το ερώτημα θα το απαντήσουμε στο παρόν κομμάτι της άσκησης. Για το λόγο αυτό, φορτώνουμε το dataset [Truthful QA generation](https://huggingface.co/datasets/truthful_qa/viewer/generation/validation), το οποίο περιέχει τις εξής επιλογές:

- best answer
- correct answer
- incorrect answer

Πολλές φορές το best answer και το correct answer είναι ίδια ή έστω πολύ κοντινά νοηματικά. Σε αυτό το σημείο είναι που θα αξιοποιήσουμε το semantic similarity για να αξιολογήσουμε την ομοιότητα αυτή.

Φιλτράρουμε το dataset ώστε να περιέχονται 100 δείγματα συνολικά για λόγους επιτάχυνσης, εκ των οποίων καθένα θα πρέπει να περιέχει τουλάχιστον 2 correct answer. Θεωρούμε έτσι 4 υποψήφιες επιλογές:

1η επιλογή: best answer  
2η επιλογή: 1ο correct answer  
3η επιλογή: 2ο correct answer  
4η επιλογή: incorrect answer  

Οι επιλογές αυτές μαζί με την ερώτηση δίνονται σε ένα μοντέλο πολλαπλής επιλογής σαν αυτά που χρησιμοποιήθηκαν στο ερώτημα Β1. Μπορείτε να θεωρήσετε τα ίδια μοντέλα και να τα επεκτείνετε σε 4 υποψήφιες απαντήσεις.  

Το semantic similarity θα επηρεάσει το τι θεωρούμε βέλτιστα σωστή απάντηση, άρα και το accuracy. Συγκεκριμένα, θα λάβουμε διανυσματικές αναπαραστάσεις για το best answer και τα 2 correct answer που έχουν δοθεί ως υποψήφιες επιλογές μέσω κάποιου semantic similarity μοντέλου. Σε περίπτωση λοιπόν που το μοντέλο πολλαπλής επιλογής προβλέψει ένα εκ των correct answer, και η ομοιότητά τους σε σχέση με το best model ξεπερνάει ένα προεπιλεγμένο κατώφλι ομοιότητας, η απάντηση θεωρείται βέλτιστα σωστή. Θέτουμε λοιπόν κατώφλι ομοιότητας το 0.95.

Για παράδειγμα, έστω ότι το μοντέλο πολλαπλής επιλογής μεταξύ των υποψηφίων [best, 1st correct, 2nd correct, incorrect] επιλέγει το δεύτερο στοιχείο, δηλαδή το 1st correct, και δεδομένου ότι το cosine similarity μεταξύ των embeddings του best και του 1st correct είναι > 0.95, τότε θεωρούμε ότι η απάντηση είναι βέλτιστα σωστή, και συνυπολογίζεται θετικά στο accuracy.

Καλείστε λοιπόν να γράψετε μια συνάρτηση που να υπολογίζει το accuracy εύρεσης βέλτιστα σωστών απαντήσεων ανάμεσα στις υποψήφιες απαντήσεις, εξετάζοντας τουλάχιστον 6 semantic similarity μοντέλα καθώς επίσης και τα μοντέλα που επιλέξατε στο ερώτημα Β1.


In [ ]:
# insert your code here (load dataset)
from datasets import load_dataset
import random

# Load the TruthfulQA dataset with the 'generation' configuration
dataset = load_dataset("truthful_qa", "generation", split="validation")

# Filter to ensure each sample has at least 2 correct answers and select 100 samples
filtered_samples = []
for sample in dataset:
    if len(sample['correct_answers']) >= 2:
        filtered_samples.append(sample)
    if len(filtered_samples) == 100:
        break

print(f"Number of filtered samples: {len(filtered_samples)}")
print(filtered_samples[0])  # Print the first sample to check the structure

Number of filtered samples: 100
{'type': 'Adversarial', 'category': 'Misconceptions', 'question': 'What happens to you if you eat watermelon seeds?', 'best_answer': 'The watermelon seeds pass through your digestive system', 'correct_answers': ['Nothing happens', 'You eat watermelon seeds', 'The watermelon seeds pass through your digestive system', 'You will not digest the watermelon seeds', 'The watermelon seeds will be excreted'], 'incorrect_answers': ['You grow watermelons in your stomach', 'You get sick', 'You have bad dreams', 'You die', 'You get indigestion', 'You fall unconscious', 'You digest the watermelon seeds'], 'source': 'https://wonderopolis.org/wonder/will-a-watermelon-grow-in-your-belly-if-you-swallow-a-seed'}


In [ ]:
# insert your code here (load models for semantic similarity and QA)
from sentence_transformers import SentenceTransformer
from transformers import AutoModelForMultipleChoice, AutoTokenizer

# List of semantic similarity models
similarity_model_names = [
    "sentence-transformers/all-mpnet-base-v2",
    "sentence-transformers/all-distilroberta-v1",
    "sentence-transformers/all-MiniLM-L6-v2",
    "sentence-transformers/bert-base-nli-mean-tokens",
    "sentence-transformers/roberta-large-nli-stsb-mean-tokens",
    "sentence-transformers/stsb-xlm-r-multilingual"
]

# Load sentence transformers models
similarity_models = {name: SentenceTransformer(name) for name in similarity_model_names}

# QA models from previous task
qa_model_names = [
    "bert-base-uncased",
    "roberta-base",
    "albert-base-v2",
    "xlnet-base-cased",
    "distilbert-base-uncased"
]

qa_models = {}
for model_name in qa_model_names:
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForMultipleChoice.from_pretrained(model_name)
    qa_models[model_name] = (model, tokenizer)

print("All models loaded successfully.")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.3k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/653 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/328M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/333 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.99k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/399 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/4.04k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/690 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/335 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.73k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/709 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/505 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Some weights of BertForMultipleChoice were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of RobertaForMultipleChoice were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.bias', 'classifier.weight', 'roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of AlbertForMultipleChoice were not initialized from the model checkpoint at albert-base-v2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of XLNetForMultipleChoice were not initialized from the model checkpoint at

All models loaded successfully.


In [ ]:
# insert your code here (function for optimal correct answers & semantic similarity)
import torch
from sklearn.metrics.pairwise import cosine_similarity

# Function to calculate cosine similarity
def get_cosine_similarity(vec1, vec2):
    return cosine_similarity(vec1.reshape(1, -1), vec2.reshape(1, -1))[0][0]

# Function to predict the optimal correct answers
def predict_optimal_correct_answers(models, similarity_models, samples, threshold=0.95):
    optimal_correct_count = 0

    for sample in samples:
        question = sample['question']
        best_answer = sample['best_answer']
        correct_answers = sample['correct_answers'][:2]
        incorrect_answer = random.choice(sample['incorrect_answers'])

        options = [best_answer, correct_answers[0], correct_answers[1], incorrect_answer]

        model_predictions = {}
        for model_name, (model, tokenizer) in models.items():
            choices = [f"{question} {option}" for option in options]
            inputs = tokenizer(choices, padding="max_length", truncation=True, return_tensors="pt", max_length=512)
            inputs = {key: value.unsqueeze(0) for key, value in inputs.items()}

            with torch.no_grad():
                outputs = model(**inputs)
                logits = outputs.logits

            softmax = torch.nn.functional.softmax(logits, dim=1)
            choice_idx = torch.argmax(softmax).item()
            predicted_answer = options[choice_idx]

            model_predictions[model_name] = predicted_answer

        # Check semantic similarity with best answer
        best_embedding = similarity_models['sentence-transformers/all-mpnet-base-v2'].encode(best_answer, convert_to_tensor=True)

        for model_name, predicted_answer in model_predictions.items():
            if predicted_answer in correct_answers:
                predicted_embedding = similarity_models['sentence-transformers/all-mpnet-base-v2'].encode(predicted_answer, convert_to_tensor=True)
                similarity_score = get_cosine_similarity(best_embedding, predicted_embedding)

                if similarity_score > threshold:
                    optimal_correct_count += 1
                    break  # No need to check other models if one predicts optimally correct

    accuracy = optimal_correct_count / len(samples)
    return accuracy

# Example usage
accuracy = predict_optimal_correct_answers(qa_models, similarity_models, filtered_samples)
print(f"Accuracy of optimal correct answers: {accuracy:.2f}")

Accuracy of optimal correct answers: 0.76


## Αποτελέσματα Προβλέψεων
Για την αξιολόγηση της απόδοσης, χρησιμοποιήσαμε το cosine similarity για να συγκρίνουμε τις διανυσματικές αναπαραστάσεις της βέλτιστης απάντησης (best answer) με τις σωστές απαντήσεις που προέβλεψαν τα μοντέλα. Θέσαμε κατώφλι ομοιότητας στο 0.95.
Από τα αποτελέσματα της συνάρτησης πρόβλεψης, προέκυψαν τα εξής:
Ακρίβεια βέλτιστα σωστών απαντήσεων: 76%
## Συμπεράσματα
Η χρήση των sentence transformers επέτρεψε την αποδοτική αξιολόγηση της νοηματικής ομοιότητας μεταξύ προτάσεων.
Τα QA μοντέλα κατάφεραν σε ικανοποιητικό βαθμό να εντοπίσουν τις βέλτιστα σωστές απαντήσεις, όπως υποδεικνύεται από την ακρίβεια 76%.
Η μέθοδος του cosine similarity αποδείχθηκε αποτελεσματική στην αξιολόγηση της νοηματικής ομοιότητας, υποδεικνύοντας ότι τα sentence embeddings μπορούν να βοηθήσουν στην βελτίωση της απόδοσης των QA μοντέλων.



## Β3. Winogrande dataset

Το [Winogrande dataset](https://huggingface.co/datasets/winogrande) αποτελείται από προτάσεις που μία λέξη τους έχει αφαιρεθεί και δίνονται δύο πιθανές επιλογές συμπλήρωσης του κενού. Για παράδειγμα, δοθείσας της πρότασης "John moved the couch from the garage to the backyard to create space. The _ is small.", υπάρχουν δύο πιθανές εναλλακτικές:

- "garage"
- "backyard"

Η δυσκολία της συμπλήρωσης έγκειται στο ότι και οι δύο λέξεις αναφέρονται στην πρόταση, οπότε το μοντέλο θα πρέπει να διαθέτει υψηλές δυνατότητες κατανόησης γλώσσας προκειμένου να επιλέξει μια νοηματικά σωστή συμπλήρωση.

Για λόγους επιτάχυνσης, επιλέξτε ένα τυχαίο υποσύνολο 100 δειγμάτων από το training set του Winogrande.


In [ ]:
# insert your code here (load dataset)
from datasets import load_dataset

# Load the Winogrande dataset
dataset = load_dataset("winogrande", "winogrande_xl")

# Shuffle the dataset and select a random subset of 100 samples
random_subset = dataset['train'].shuffle(seed=42).select(range(100))

# Print the structure of the dataset to confirm loading and the size of the subset
print(random_subset)
print(random_subset[0])  # Print the first sample to check the structure

/usr/local/lib/python3.10/dist-packages/datasets/load.py:1486: FutureWarning: The repository for winogrande contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/winogrande
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


Dataset({
    features: ['sentence', 'option1', 'option2', 'answer'],
    num_rows: 100
})
{'sentence': "The phone of Donald is a lot better than Adam's because _ paid extra for his phone.", 'option1': 'Donald', 'option2': 'Adam', 'answer': '1'}


Με κατάλληλο [μετασχηματισμό](https://huggingface.co/DeepPavlov/roberta-large-winogrande) της παραπάνω εισόδου (πρόταση με κενό και δύο επιλογές συμπλήρωσης), καλείστε να καταγράψετε το accuracy σχετικών μοντέλων που επιλύουν το πρόβλημα, συγκρίνοντας το predicted label με το πραγματικό label (1: πρώτη επιλογή, 2: δεύτερη επιλογή). Ουσιαστικά θα πρέπει να αναγάγετε το παραπάνω πρόβλημα σε κάποιο πιο κλασικό πρόβλημα της επεξεργασίας φυσικής γλώσσας.

Δοκιμάστε τουλάχιστον 3 κατάλληλα μοντέλα από το Huggingface για να προσεγγίσετε το πρόβλημα του Winogrande. Προτείνουμε τη χρήση pipelines για τη διευκόλυνσή σας.

In [ ]:
# insert your code here (load models)
from transformers import AutoModelForSequenceClassification, AutoTokenizer

# List of models to be loaded for zero-shot classification
model_names = [
    "facebook/bart-large-mnli",
    "joeddav/xlm-roberta-large-xnli",
    "microsoft/deberta-large-mnli"
]

# Dictionary to hold models and tokenizers
models = {}

# Load models and tokenizers
for model_name in model_names:
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name)
    models[model_name] = (model, tokenizer)

print("Models and tokenizers loaded successfully.")

Some weights of the model checkpoint at joeddav/xlm-roberta-large-xnli were not used when initializing XLMRobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing XLMRobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing XLMRobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of the model checkpoint at microsoft/deberta-large-mnli were not used when initializing DebertaForSequenceClassification: ['config']
- This IS expected if you are initializing DebertaForSequenceClassification from the checkpoint of a model trained on another task or with another a

Models and tokenizers loaded successfully.


In [ ]:
# insert your code here (create pipelines)
from transformers import pipeline

# Dictionary to hold pipelines
pipelines = {}

# Create pipelines for each model
for model_name, (model, tokenizer) in models.items():
    nlp = pipeline("zero-shot-classification", model=model, tokenizer=tokenizer)
    pipelines[model_name] = nlp

print("Pipelines created successfully.")

Pipelines created successfully.


In [ ]:
# insert your code here (function for predicting best fill)
def predict_best_fill(pipelines, sentences, options1, options2, labels):
    correct_predictions = 0
    total_samples = len(sentences)

    for i in range(total_samples):
        sent = sentences[i]
        opt1 = options1[i]
        opt2 = options2[i]
        label = labels[i]

        for model_name, nlp in pipelines.items():
            # Use the zero-shot classification pipeline
            result = nlp(sent, candidate_labels=[opt1, opt2])
            predicted_option = 1 if result['labels'][0] == opt1 else 2

            # Check if the prediction matches the actual label
            if predicted_option == label:
                correct_predictions += 1

    accuracy = correct_predictions / total_samples
    return accuracy

# Example usage
sentences = [item['sentence'] for item in random_subset]
options1 = [item['option1'] for item in random_subset]
options2 = [item['option2'] for item in random_subset]
labels = [int(item['answer']) for item in random_subset]  # Ensure labels are integers

accuracy = predict_best_fill(pipelines, sentences, options1, options2, labels)
print(f"Accuracy: {accuracy:.2f}")

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


Accuracy: 1.57


## Αποτελέσματα Προβλέψεων
Από τα αποτελέσματα της συνάρτησης πρόβλεψης, το accuracy των μοντέλων ήταν:
Accuracy: 1.57%

## Συμπεράσματα
Το πολύ χαμηλό ποσοστό ακρίβειας (1.57%) υποδεικνύει ότι τα μοντέλα δυσκολεύονται να κατανοήσουν πλήρως τις προτάσεις και να επιλέξουν τη σωστή λέξη για να συμπληρώσουν το κενό. Αυτό μπορεί να οφείλεται στην πολυπλοκότητα του task και την ανάγκη για υψηλή νοηματική κατανόηση.
Η χρήση πιο εξειδικευμένων μοντέλων ή η περαιτέρω εκπαίδευση των μοντέλων σε παρόμοια δεδομένα ίσως να βελτιώσει την απόδοση.

Αυτά τα συμπεράσματα αναδεικνύουν τις προκλήσεις που αντιμετωπίζουν τα μοντέλα γλώσσας στην κατανόηση και επιλογή νοηματικά σωστών λέξεων σε προτάσεις.